File: src/conventions/teaching/parse-tree-builder.ts
/**
 * Parse Tree Builder
 *
 * Builds a ParseTreeView from arbitration results and decision provenance.
 * Shows the full decision chain: which convention modules were considered,
 * why each was accepted/rejected, and the path to the correct bid.
 *
 * Pure function — no side effects, no imports from strategy/stores/components.
 */

import type { ArbitrationResult } from "../pipeline/pipeline-types";
import type { DecisionProvenance } from "../pipeline/evaluation/provenance";
import type { Call } from "../../engine/types";
import type {
  ParseTreeView,
  ParseTreeModuleNode,
  ParseTreeModuleVerdict,
  ParseTreeCondition,
} from "./teaching-types";
import type { CatalogIndex } from "./teaching-projection-builder";
import { resolveDisplayText } from "./teaching-projection-builder";

/** Build a ParseTreeView from arbitration results and provenance.
 *  Groups proposals by module, determines verdict per module,
 *  and constructs the decision chain. */
export function buildParseTree(
  arbitration: ArbitrationResult,
  provenance: DecisionProvenance,
  catalogIndex?: CatalogIndex,
): ParseTreeView {
  const selectedModuleId = arbitration.selected?.proposal.moduleId;

  // Collect per-module data: meanings proposed, verdicts, conditions
  const moduleMap = new Map<string, {
    displayLabel: string;
    truthMeanings: { meaningId: string; displayLabel: string; call?: Call }[];
    eliminatedMeanings: { meaningId: string; displayLabel: string; reason?: string }[];
    conditions: ParseTreeCondition[];
    eliminationReason?: string;
  }>();

  // Process truth-set entries (modules that produced valid proposals)
  for (const encoded of arbitration.truthSet) {
    const moduleId = encoded.proposal.moduleId;
    const entry = getOrCreate(moduleMap, moduleId);
    if (!encoded.proposal.teachingLabel && import.meta.env.DEV) {
      console.warn(`[teaching] No teachingLabel for meaningId: ${encoded.proposal.meaningId}`);
    }
    entry.truthMeanings.push({
      meaningId: encoded.proposal.meaningId,
      displayLabel: encoded.proposal.teachingLabel?.name ?? encoded.proposal.meaningId,
      call: encoded.call,
    });
    // Collect conditions from the first truth-set meaning per module
    if (entry.conditions.length === 0) {
      entry.conditions = encoded.proposal.clauses.map(clause => ({
        factId: clause.factId,
        description: resolveDisplayText(catalogIndex?.byFactId.get(clause.factId))
          ?? clause.description,
        satisfied: clause.satisfied,
        observedValue: clause.observedValue,
      }));
    }
  }

  // Process eliminations (modules that proposed but were rejected)
  for (const elimination of arbitration.eliminations) {
    const moduleId = elimination.moduleId;
    const entry = getOrCreate(moduleMap, moduleId);
    entry.eliminatedMeanings.push({
      meaningId: elimination.candidateBidName,
      displayLabel: elimination.candidateBidName,
      reason: elimination.reason,
    });
    if (!entry.eliminationReason) {
      entry.eliminationReason = elimination.reason;
    }
  }

  // Enrich eliminated modules with conditions from provenance
  for (const elimTrace of provenance.eliminations) {
    // Find which module this elimination belongs to
    for (const [, data] of moduleMap) {
      const hasElim = data.eliminatedMeanings.some(
        m => m.meaningId === elimTrace.candidateId,
      );
      if (hasElim && data.conditions.length === 0 && elimTrace.evidence.length > 0) {
        data.conditions = elimTrace.evidence.map(ev => {
          const catalogText = resolveDisplayText(catalogIndex?.byFactId.get(ev.factId ?? ev.conditionId));
          if (!catalogText && !ev.description && catalogIndex && import.meta.env.DEV) {
            console.warn(`[teaching] No catalog entry found for factId: ${ev.factId ?? ev.conditionId}`);
          }
          return {
            factId: ev.conditionId,
            description: catalogText ?? ev.description ?? ev.conditionId,
            satisfied: ev.satisfied,
            observedValue: ev.observedValue,
          };
        });
      }
    }
  }

  // Enrich with activation traces (modules that were activated/deactivated)
  for (const activation of provenance.activation) {
    const entry = getOrCreate(moduleMap, activation.moduleId);
    if (!activation.activated && !entry.eliminationReason) {
      entry.eliminationReason = activation.reason ?? "Module not activated";
    }
  }

  // Build module nodes
  const modules: ParseTreeModuleNode[] = [];
  for (const [moduleId, data] of moduleMap) {
    const isSelected = moduleId === selectedModuleId;
    const hasTruthMeanings = data.truthMeanings.length > 0;

    let verdict: ParseTreeModuleVerdict;
    if (isSelected) {
      verdict = "selected";
    } else if (hasTruthMeanings) {
      verdict = "applicable";
    } else {
      verdict = "eliminated";
    }

    const meanings = [
      ...data.truthMeanings.map(m => ({
        meaningId: m.meaningId,
        displayLabel: m.displayLabel,
        matched: true as const,
        call: m.call,
      })),
      ...data.eliminatedMeanings.map(m => ({
        meaningId: m.meaningId,
        displayLabel: m.displayLabel,
        matched: false as const,
      })),
    ];

    modules.push({
      moduleId,
      displayLabel: data.displayLabel,
      verdict,
      conditions: data.conditions,
      meanings,
      ...(verdict === "eliminated" && data.eliminationReason
        ? { eliminationReason: data.eliminationReason }
        : {}),
    });
  }

  // Sort: selected first, then applicable, then eliminated
  const verdictOrder: Record<ParseTreeModuleVerdict, number> = {
    selected: 0,
    applicable: 1,
    eliminated: 2,
  };
  modules.sort((a, b) => verdictOrder[a.verdict] - verdictOrder[b.verdict]);

  // Build selected path
  const selectedPath = arbitration.selected
    ? {
        moduleId: arbitration.selected.proposal.moduleId,
        meaningId: arbitration.selected.proposal.meaningId,
        call: arbitration.selected.call,
      }
    : null;

  return { modules, selectedPath };
}

function getOrCreate(
  map: Map<string, {
    displayLabel: string;
    truthMeanings: { meaningId: string; displayLabel: string; call?: Call }[];
    eliminatedMeanings: { meaningId: string; displayLabel: string; reason?: string }[];
    conditions: ParseTreeCondition[];
    eliminationReason?: string;
  }>,
  moduleId: string,
) {
  let entry = map.get(moduleId);
  if (!entry) {
    entry = {
      displayLabel: moduleId,
      truthMeanings: [],
      eliminatedMeanings: [],
      conditions: [],
    };
    map.set(moduleId, entry);
  }
  return entry;
}

File: src-tauri/crates/bridge-conventions/src/teaching/parse_tree_builder.rs
//! Parse tree builder — builds ParseTreeView from PipelineResult.
//!
//! Mirrors TS from `conventions/teaching/parse-tree-builder.ts`.

use std::collections::HashMap;

use crate::pipeline::types::PipelineResult;
use crate::teaching::teaching_types::*;

/// Build a parse tree view from a PipelineResult.
pub fn build_parse_tree(result: &PipelineResult) -> ParseTreeView {
    let mut module_nodes: HashMap<String, ParseTreeModuleNode> = HashMap::new();

    // Process truth set carriers (selected + applicable)
    for carrier in &result.truth_set {
        let module_id = &carrier.proposal().module_id;
        let node = module_nodes.entry(module_id.clone()).or_insert_with(|| {
            ParseTreeModuleNode {
                module_id: module_id.clone(),
                display_label: module_id.clone(),
                verdict: ParseTreeModuleVerdict::Applicable,
                conditions: Vec::new(),
                meanings: Vec::new(),
                elimination_reason: None,
            }
        });

        node.meanings.push(ParseTreeMeaning {
            meaning_id: carrier.proposal().meaning_id.clone(),
            display_label: carrier.proposal().teaching_label.name.to_string(),
            matched: carrier.encoded.eligibility.hand.satisfied,
            call: Some(carrier.call().clone()),
        });
    }

    // Process eliminated carriers
    for carrier in &result.eliminated {
        let module_id = &carrier.proposal().module_id;
        let node = module_nodes.entry(module_id.clone()).or_insert_with(|| {
            ParseTreeModuleNode {
                module_id: module_id.clone(),
                display_label: module_id.clone(),
                verdict: ParseTreeModuleVerdict::Eliminated,
                conditions: Vec::new(),
                meanings: Vec::new(),
                elimination_reason: carrier
                    .traces
                    .elimination
                    .as_ref()
                    .map(|e| e.reason.clone()),
            }
        });

        // Add conditions from failed clauses
        for clause in &carrier.proposal().clauses {
            if !clause.satisfied {
                node.conditions.push(ParseTreeCondition {
                    fact_id: clause.fact_id.clone(),
                    description: clause
                        .description
                        .clone()
                        .unwrap_or_else(|| clause.fact_id.clone()),
                    satisfied: false,
                    observed_value: clause.observed_value.clone(),
                });
            }
        }

        node.meanings.push(ParseTreeMeaning {
            meaning_id: carrier.proposal().meaning_id.clone(),
            display_label: carrier.proposal().teaching_label.name.to_string(),
            matched: false,
            call: Some(carrier.call().clone()),
        });
    }

    // Mark selected module
    if let Some(ref selected) = result.selected {
        if let Some(node) = module_nodes.get_mut(&selected.proposal().module_id) {
            node.verdict = ParseTreeModuleVerdict::Selected;
        }
    }

    // Build selected path
    let selected_path = result.selected.as_ref().map(|s| SelectedPath {
        module_id: s.proposal().module_id.clone(),
        meaning_id: s.proposal().meaning_id.clone(),
        call: s.call().clone(),
    });

    // Sort modules: selected first, then applicable, then eliminated
    let mut modules: Vec<ParseTreeModuleNode> = module_nodes.into_values().collect();
    modules.sort_by_key(|m| match m.verdict {
        ParseTreeModuleVerdict::Selected => 0,
        ParseTreeModuleVerdict::Applicable => 1,
        ParseTreeModuleVerdict::Eliminated => 2,
    });

    ParseTreeView {
        modules,
        selected_path,
    }
}
